Add policies readme (#400)
netgroup-polito/CrownLabs: policies/README.md

## OPA POLICIES

Policies are rules that can be applied when any of your cluster object is _created_, _updated_, _patched_ and _deleted_.

Policies are always executed _before_ the action takes place and their goal is to allow or prevent the requested actions.

> The language used to express OPA policies is [REGO] (https://www.openpolicyagent.org/docs/latest/policy-language/).

Reviewer: ```suggestion
The language used to express OPA policies is [REGO](https://www.openpolicyagent.org/docs/latest/policy-language/).
```

Otherwise the link is not displayed correctly

### HOW TO USE

To use OPA gatekeeper you need to:

- Install Gatekeeper in your cluster, deploying the CRDs and the associated operators/components. See [this README](../infrastructure/policy-enforcement/README.md) for further information about the installation process.

- Create your own constraint and template files and, if needed, the config-sync file. See [the examples folder](../infrastructure/policy-enforcement/examples) for an example of the three resources.

Now you can test your policies on your cluster objects.

### HOW TO USE GATEKEEPER

### Template

In the Template you should define your own constraint CRD schema and associate it to your rego policies, such as in this example: [**template.yaml**](../infrastructure/policy-enforcement/examples/template.yaml).

> Under the _crd_ tag you have to define your connstraint CRD schema:

Reviewer: ```suggestion
Under the _crd_ tag you have to define your constraint CRD schema:
```

- kind: must be the same as the template name in camelcase (mandatory)
- properties: must declare your fixed parameters (optional)

> Under the _targets_ tag you have to define your rego policy.

Reviewer: I suppose a new line is missing, as it is displayed incorrectly

The example REGO policy executes the following operations:

> - the provided variable retrieves all the label keys of the input object

Reviewer: Please, use the appropriate formatting for code. E.g.:
```suggestion
- the `provided` variable retrieves all the label keys of the input object
```

- if missing is empty then count(missing)>0 will be false, the execution will be ended and so there will be no violation
- if missing is not empty then the condition will be true, the execution will continue and so there will be a violation
- the msg variable, containg the sprintf value, will be the returned value of the violation together with the details object

As you could see, you can retrieve your input object using `input` variable.
In OPA, input is a reserved, global variable whose value is the Kubernetes AdmissionReview object that the API server hands to any admission control webhook.

### Constraint

In Constraint you should declare the resources on which your Template operates and define possible optional parameters (e.g., reject a modification if a value is greater than a given threshold), for example: [**constraint.yaml**](../infrastructure/policy-enforcement/examples/constraint.yaml).
The _kind_ tag must be the one specified in the _kind_ tag of the related template.
The _match_ tag must list the cluster resources on which you want to enforce the policy.
The _paramters_ tag must list your fixed parameters, if any.

### Config-Sync

> [^1] Some constraints cannot be written without having access to more cluster resources than just the object under test. For example, it is impossible to know if an instance is valid unless a rule has access to all the templates. To make such rules possible, we enable syncing of data into OPA.

Reviewer: ```suggestion
Some constraints cannot be written without having access to more cluster resources than just the object under test. For example, it is impossible to know if an instance is valid unless a rule has access to all the templates. To make such rules possible, we enable syncing of data into OPA. [1]
```

Feel more standard

The OPA rules do not require cluster resources by default, however if you need them you must first cache your objects before using them in your contraint violations. 

> The Config resource has to be named "config" for it to be reconciled by Gatekeeper. Gatekeeper will ignore the resource if you do not name it "config".

Reviewer: ```suggestion
**Note:** The Config resource has to be named `config` for it to be reconciled by Gatekeeper. Gatekeeper will ignore the resource if you do not name it `config`.
```

Additionally, you never introduced what the Config resource is.

### How to write rego policies in gatekeeper
You should use `violation` functions to write your rules.

    violation[{"msg":msg,"details":{}}]{
    }
    
 `msg` and `details`are the values returned if a violation occurs, otherwise nothing is returned.
The rule must have at least one boolean condition to take a decision whether the violation occurs or not.
If you need to specify more boolean conditions, proceed as follows:

 - For `and` conditions (all your conditions must be true in order for the violation to take place):
    write all of them in one violation function
  
 - For `or` conditions (at least one condition must be true in order for the violation to take place)
   write one violation function for each condition
  
In general OPA policies in gatekeeper work as black list mode: so you specify what can't be done.
 You can learn [REGO](https://www.openpolicyagent.org/docs/latest/#rego) language.
 You can exploit [PLAYGROUND](https://play.openpolicyagent.org/) to try your rules.

### Note
[^1] This information is taken from the official [doucumentation](https://open-policy-agent.github.io/gatekeeper/website/docs/sync).

> [^2] This information is taken from the official [doucumentation](https://www.openpolicyagent.org/docs/latest/kubernetes-introduction/#how-does-it-work-with-plain-opa-and-kube-mgmt).

Reviewer: A new line is missing. There is a typo in the word `documentation`.

# OPA policies

> Policies are rules that can be applied when any of your cluster object is _createdn_, _updated_, _patched_ and _deleted_.

Reviewer 2: ```suggestion
Policies are rules that can be applied when any of your cluster object is _created_, _updated_, _patched_ and _deleted_.
```

Policies are always executed _before_ the action takes place and their goal is to allow or prevent the requested modification.

This solution is very fast in term of required computing time thanks to the efficienty of the OPA gatekeeper.

## How to use policies
To use OPA gatekeeper you need to:

> - Deploy a Release using Prebuilt Image. See [link](../infrastructure/policy-enforcement/README.md).

Reviewer 2: Not clear to me. What did you want to say? That you need to have a gatekeeper running in your cluster, so you have to depoy it?

> - Create your own constraint, template and eventually config-sync. See [examples](../infrastructure/policy-enforcement/examples).

Reviewer 2: eventually = "finalmente". Is it that what you wanted to say?

Now you can test your policies on your cluster objects.

## How to use the Gatekeeper

### Template

> In the Template you should define your own constraint CRD and associate it to your rego policies, such as in this [**template.yaml**](../infrastructure/policy-enforcement/examples/template.yaml) example.

Reviewer 2: "rego" was never mentioned before. Reader has no idea what this is.
Possible rephrasing: ... to your rego (i.e., the language used by the OPA gatekeeper) policies....

### Constraint

> In Constraint you should declare on which resources your Template must be enforced and define your parameters values, such as in this [**constraint.yaml**](../infrastructure/policy-enforcement/examples/constraint.yaml) example.

Reviewer 2: Possible rephrasing:
In Constraint you should declare the resources on which your Template operates on and define possible optional parameters (e.g., reject a modification if a value is greater than a given threshold)

> You can retrieve your input object using `input` variable.

Reviewer 2: This sentence is really mysterious to me. No idea what is the input variable.

### Config-Sync

> If you need to get resources from the cluster to write your policy, you need first to cache those and then you will be able to retrieve them in the Template rules using `data.inventory`, such as in this [**config-sync.yaml**](../infrastructure/policy-enforcement/examples/config-sync.yaml) example.

Reviewer 2: What does it mean "cache" a resource? How can I "cache" a resource?

### Retrieve cached resources

> To retrieve the cached resources in the rego rules you can do as follow:

Reviewer 2: Provide some more context. For a newby reader, those variables mean almost nothing and he has no idea what to do with them.

- For cluster-scoped objects: `data.inventory.cluster[<groupVersion>][<kind>][<name>]`
- Example referencing the Gatekeeper namespace: `data.inventory.cluster["v1"].Namespace["gatekeeper"]`
- For namespace-scoped objects: `data.inventory.namespace[<namespace>][groupVersion][<kind>][<name>]`
- Example referencing the Gatekeeper pod: `data.inventory.namespace["gatekeeper"]["v1"]["Pod"]["gatekeeper-controller-manager-d4c98b788-j7d92"]`

### Retrieve input object

> The  `input`  document contains the following fields:

Reviewer 2: ```suggestion
The `input` document contains the following fields:
```

- `input.review.kind` specifies the type of the object (e.g., `Pod`, `Service`, etc.)
- `input.review.operation` specifies the type of the operation, i.e., `CREATE`, `UPDATE`, `DELETE`, `CONNECT`.
- `input.review.userInfo` specifies the identity of the caller.
- `input.review.object` contains the entire Kubernetes object.
- `input.review.oldObject` specifies the previous version of the Kubernetes object on `UPDATE` and `DELETE`.

### How to write rego policies in gatekeeper

> You should use `violation` functions to write you rules:

Reviewer 2: ```suggestion
You should use `violation` functions to write your rules:
```

## OPA POLICIES

Policies are rules that can be applied at creation, update, patch and deletion time to any of your cluster objects.

> Policies are always executed before the action takes place and its goal is to allow or prevent the modification.

Reviewer 2: ```suggestion
Policies are always executed _before_ the action takes place and their goal is to allow or prevent the modification.
```

> This solution is much performing in terms of execution time because it exploits OPA gatekeeper.

Reviewer 2: ```suggestion
This solution is very efficient in term of required computing time thanks to the OPA gatekeeper.
```

### HOW TO USE

To use OPA gatekeeper you need to:

> - Deploy a Release using Prebuilt Image. See [link](../infrastructure/policy-enforcement/README.md)

Reviewer 2: ```suggestion
- Deploy a Release using Prebuilt Image. See [link](../infrastructure/policy-enforcement/README.md).
```

> - Create your own constraint, template and eventually config-sync. See [examples](../infrastructure/policy-enforcement/examples)

Reviewer 2: ```suggestion
- Create your own constraint, template and eventually config-sync. See [examples](../infrastructure/policy-enforcement/examples).
```

Now you can test your policies on your cluster objects.

### HOW TO USE GATEKEEPER

### Template

> In the Template you should define your own constraint CRD and associate it to your rego policies, for example : [ **template.yaml**](../infrastructure/policy-enforcement/examples/template.yaml)

Reviewer 2: ```suggestion
In the Template you should define your own constraint CRD and associate it to your rego policies, such as in this example: [**template.yaml**](../infrastructure/policy-enforcement/examples/template.yaml)
```

- For namespace-scoped objects: `data.inventory.namespace[<namespace>][groupVersion][<kind>][<name>]`

- Example referencing the Gatekeeper pod: `data.inventory.namespace["gatekeeper"]["v1"]["Pod"]["gatekeeper-controller-manager-d4c98b788-j7d92"]`

### Retrieve input object

The  `input`  document contains the following fields:

-   `input.review.kind`  specifies the type of the object (e.g.,  `Pod`,  `Service`, etc.)
-   `input.review.operation`  specifies the type of the operation, i.e.,  `CREATE`,  `UPDATE`,  `DELETE`,  `CONNECT`.
-   `input.review.userInfo`  specifies the identity of the caller.
-   `input.review.object`  contains the entire Kubernetes object.
-   `input.review.oldObject`  specifies the previous version of the Kubernetes object on  `UPDATE`  and  `DELETE`.

### How to write rego policies in gatekeeper
You should use `violation` functions to write you rules.

    violation[{"msg":msg,"details":{}}]{
    }
    
 `msg` and `details`are the values returned if a violation occurs, otherwise nothing is returned.
The rule must have at least one boolean condition to take a decision wether the violation occurs or not.

> If you need to specify more boolean conditions do as follow:

Reviewer 2: ```suggestion
If you need to specify more boolean conditions, proceed as follows:
```

>  - For in `and` conditions (all your conditions must be true in order for the violation to take place):

Reviewer 2: ```suggestion
 - For `and` conditions (all your conditions must be true in order for the violation to take place):
```

    write all of them in one violation function
  

>  - For in `or` conditions (at least one condition must be true in order for the violation to take place)

Reviewer 2: ```suggestion
 - For `or` conditions (at least one condition must be true in order for the violation to take place)
```

## OPA POLICIES

Policies are rules that can be applied when any of your cluster object is _created_, _updated_, _patched_ and _deleted_.

> Policies are always executed _before_ the action takes place and their goal is to allow or prevent the modification.

Reviewer 2: ```suggestion
Policies are always executed _before_ the action takes place and their goal is to allow or prevent the requested action.
```

- Deploy a Release using Prebuilt Image to create in your cluster the custom objects needed for gatekeeper to work. See [link](../infrastructure/policy-enforcement/README.md).

- Create your own constraint and template files and eventually, if needed, config-sync file. See [examples](../infrastructure/policy-enforcement/examples).

Now you can test your policies on your cluster objects.

### HOW TO USE GATEKEEPER

### Template

In the Template you should define your own constraint CRD and associate it to your rego policies, such as in this example: [**template.yaml**](../infrastructure/policy-enforcement/examples/template.yaml)

### Constraint

In Constraint you should declare the resources on which your Template operates and define possible optional parameters (e.g., reject a modification if a value is greater than a given threshold), for example: **[constraint.yaml](../infrastructure/policy-enforcement/examples/constraint.yaml)**
You can retrieve your input object using `input` variable.
In OPA, input is a reserved, global variable whose value is the Kubernetes AdmissionReview object that the API server hands to any admission control webhook.

### Config-Sync

> Some constraints are impossible to write without access to more state than just the object under test. For example, it is impossible to know if an instance is valid unless a rule has access to all the templates. To make such rules possible, we enable syncing of data into OPA.

Reviewer: Again, a reference should be added if you copy text from other places, such as the official documentation.

> Some constraints are impossible to write without access to more state than just the object under test. For example, it is impossible to know if an instance is valid unless a rule has access to all the templates. To make such rules possible, we enable syncing of data into OPA.

Reviewer 2: ```suggestion
Some constraints cannot be written without having access to more state than just the object under test. For example, it is impossible to know if an instance is valid unless a rule has access to all the templates. To make such rules possible, we enable syncing of data into OPA.
```
Tried to rephrase, but still unsure what the first sentence means ("to more state": what does this mean?)

The OPA rules do not require cluster resources by default, however if you need them you must first cache your objects before using them in your contraint violations. 

Kubernetes data can be replicated into OPA via the sync config resource. Currently resources defined in `syncOnly` will be synced into OPA. Updating syncOnly should dynamically update what objects are synced.

> Once data is synced into OPA, rules can access the cached data under the `data.inventory` document, for example: **[config-sync.yaml](../infrastructure/policy-enforcement/examples/config-sync.yaml)**

Reviewer: I would add a note that the config-sync must be named 'config', otherwise it is ignored by gatekeeper (https://open-policy-agent.github.io/gatekeeper/website/docs/sync). This is something I initially missed when trying to deploy the policies in the cluster.

- Example referencing the Gatekeeper namespace: `data.inventory.cluster["v1"].Namespace["gatekeeper"]`

- For namespace-scoped objects: `data.inventory.namespace[<namespace>][groupVersion][<kind>][<name>]`

- Example referencing the Gatekeeper pod: `data.inventory.namespace["gatekeeper"]["v1"]["Pod"]["gatekeeper-controller-manager-d4c98b788-j7d92"]`

### Retrieve input object

The `input` document contains the following fields:

-   `input.review.kind`  specifies the type of the object (e.g.,  `Pod`,  `Service`, etc.)
-   `input.review.operation`  specifies the type of the operation, i.e.,  `CREATE`,  `UPDATE`,  `DELETE`,  `CONNECT`.
-   `input.review.userInfo`  specifies the identity of the caller.
-   `input.review.object`  contains the entire Kubernetes object.
-   `input.review.oldObject`  specifies the previous version of the Kubernetes object on  `UPDATE`  and  `DELETE`.

### How to write rego policies in gatekeeper
You should use `violation` functions to write your rules.

    violation[{"msg":msg,"details":{}}]{
    }
    
 `msg` and `details`are the values returned if a violation occurs, otherwise nothing is returned.

> The rule must have at least one boolean condition to take a decision wether the violation occurs or not.

Reviewer: ```suggestion
The rule must have at least one boolean condition to take a decision whether the violation occurs or not.
```
